flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -162,7 +174,14 @@ public Set<Long> unregisterUnusedState(long lowestCheckpointID) {
                         subsumed.add(entry.stateHandle);
                     }
                     it.remove();
-                } else {
+                } else if (preventsDiscardingCreatedCheckpoint(entry)) {
```

Comment:
`lowestCheckpointID` is still valid because it is not subsumed. It is `CheckpointSubsumeHelper` and checkpoint store that are responsible for subsumption.
 
This method, however, returns potentially subsumed checkpoints but whose state might be in use by other checkpoints.
I don't think these responsibilites should be mixed.

Besides that, this seems unrelated to this PR.